### DISCUSSION TITLE
Highlight exports: Memos / Nextcloud, etc...?

### DISCUSSION THREAD

**Discussion by abdessalaam** [koreader/koreader#10923] [Q&A]
### Memos
Has anyone managed to export highlights to Memos? I have generated API, and set it in settings.reader.lua, I have later also tried replacing API with curl command (following [Memos docs](https://usememos.com/docs/access-tokens)):

` curl https://demo.usememos.com/api/v1/memo  -H "Accept: application/json" -H "Authorization: Bearer {YOUR_ACCESS_TOKEN}" `

but even though the message says export to memos was successful, nothing shows in Memos....

### Other exports
On a slightly different note, would it be possible to export highlights to Notion, Nextcloud (notes, or as files)...?

Thanks for any advice!

**Comment by pazos** [comment]
Does the API key shows up in the setup menu for memos (in KOReader)?. If not please introduce a bogus value in there, save and, on your computer, check the proper name of the setting".

I'm speaking from memory so I'm not sure if the settings for the exporter plugin are stored in `settings.reader.lua` or elsewhere.

**Reply by abdessalaam** [reply]
Thanks. Yes, the api shows there! I am puzzling over where to specify my Memos installation address? That's why I added the whole curl command, but it didn't help.

**Reply by pazos** [reply]
It looks that this particular target isn't well documented.
The api key should include the FQDN of the server that hosts the service. Just make a string that you can copy & paste on your browser. Make sure it works there and then use the same string on the API field of the plugin.

**Reply by abdessalaam** [reply]
I think I will give up - this is way out of my depth. I tried various ways incl. ` https://my.memos.tld/api/v1/memo {MY_API_TOKEN}
 ` and the message still says "export successful", but nothing shows in my Memos... 😭

**Reply by pazos** [reply]
pinging @fierceX, who implemented that particular exporter.

**Reply by abdessalaam** [reply]
Thank you! If @fierceX wouldn't mind sharing how to configure it correctly, I'd be really grateful!

**Reply by lbesnard** [reply]
I'm in the same boat and would appreciate a bit more of doco to set this up correctly

**Reply by pazos** [reply]
If somebody of you can share with me a temporal url and token I can give it a try. In that case send me an email.

Otherwise we will need to wait for the original author or for a new maintainer.

If `memos` was not free software I would gladly remove its exporter, given it isn't well documented.

**Reply by abdessalaam** [reply]
Thanks @pazos — what's the best way to contact you? I have created access for you.

**Reply by pazos** [reply]
ap, sorry. I was under the impression my email was public at my github profile.

Please send me an email to paziusss at gmail dot com

**Reply by abdessalaam** [reply]
thank you — done!

**Comment by pazos** [comment]
> On a slightly different note, would it be possible to export highlights to Notion, Nextcloud (notes, or as files)...?

Would it be possible? Yes.

The exporter plugin should be able to export to any kind of file or service given that the format or protocol is known and it isn't a pain to implement.

But at the end of the day it really depends on volunteers to write the specific exporters, submit a patch and maintain such target. So probably we won't be able unless somebody steps in ;)

**Reply by abdessalaam** [reply]
Had I known how, I would have stepped in :)

**Comment by fierceX** [comment]
@bpawel-bclub @pazos 

Sorry, I ignored the message on github because of the Chinese holiday. Now I found out that open_api has been removed on open source memos, causing the current plugin to fail. But before that, the project used the same open_api model as flomo, a Chinese piece of business note-taking app. Now these two services have changed, so I'll modify that memos plugin to the new version first, and then add a new version of the old flomo, do you think this will work?

**Reply by Explorare** [reply]
Sounds great. I also dig this issue yesterday and found out that the openAPI has been deprecated recently. Failed to fork the plugin to meet the new API. Looking forward to your PR.

**Reply by fierceX** [reply]
I've filed a PR to fix this.：https://github.com/koreader/koreader/pull/10988

**Reply by Explorare** [reply]
The patch for Memos works like a charm. Thank you for your effort.
And the API URL that works for me is `http://192.168.0.11:5230/api/v1/memo`. The default token has a 7 day expire limit so I generated a new one.

**Reply by abdessalaam** [reply]
Amazing, thank you @fierceX — your patch works beautifully! Really appreciate your time and commitment.

**Reply by pazos** [reply]
Awesome, thank you!!!

@fierceX, @Explorare, @bpawel-bclub: could you please document the setup in https://github.com/koreader/koreader/wiki/Highlight-export ??

**Reply by abdessalaam** [reply]
I have [made an edit](https://github.com/koreader/koreader/wiki/Highlight-export), please feel free to amend / correct.

**Reply by abdessalaam** [reply]
> I've filed a PR to fix this.：https://github.com/koreader/koreader/pull/10988

A quick additional question @fierceX (sorry if it should be in a separate thread?): would it be possible to replace spaces with underscores or hyphens in the automatically added hashtags? Currently only the first word of the title becomes the tag, and it's often "The" 😉

**Reply by abdessalaam** [reply]
This could be lines 138 — 142 in memos.lua:
```
-- Replace spaces with underscores in booknotes.title
local title_with_underscores = string.gsub(booknotes.title, " ", "_")
-- Create the highlight hashtag with the modified title
highlight =  highlight .. booknotes.title .. " (page: " .. clipping.page .. "）\n\n #" .. title_with_underscores .. " #koreader"
local result, err = makeRequest("POST", { content = highlight }, self.settings.api, self.settings.token)
```

**Reply by fierceX** [reply]
@bpawel-bclub Sorry, I didn't take into account the English title, I'll modify the code

**Reply by abdessalaam** [reply]
Thank you, the lines I suggested above work well for me.

**Reply by Amolith** [reply]
I believe this plugin _may_ be broken again, as I'm experiencing identical behaviour to the OP. I have highlights in a book, press the export button, get the popup saying everything has been successfully exported, check Memos, and it's still completely empty. The API URL is set, the token is set, and `Export to Memos` is checked as well.

**Reply by fierceX** [reply]
@Amolith Can you tell me what version of memos you are using, I'll check and test it sometime

**Reply by Amolith** [reply]
@fierceX at the moment, I'm one release behind, on `v0.19.0`.

**Reply by fierceX** [reply]
@Amolith I checked on my side, it works, first, you have to set the api address, remember to add the suffix `/api/v2/memos` to your url, for example the full one is: `https:memos.example.com/api/v2/memos`. Later, I'll add a check to see if it was created successfully
